src/models/generalized_mixture_reach_surface.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class for modeling Pairwise union reach surface."""
+
+import copy
+from typing import List
+import numpy as np
+from cvxopt import matrix
+from cvxopt import solvers
+from typing import Iterable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_surface import ReachSurface
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+solvers.options["show_progress"] = False
+
+
+class GeneralizedMixtureReachSurface(ReachSurface):
+    """Models reach with the generalized mixture overlap model.
+
+    GeneralizedMixtureReachSurface models the combined reach with the formula
+
+
+    N * sum_{j=1..c} [1 - prod_{i=1..p} (1-a_i_j r_i / N)] +epsilon
+
+    where :
+          all a_i_j >= 0
+          sum_{j=1..c} a_i_j=1 for all i.
+
+    """
+
+    def __init__(
+        self,
+        reach_curves: Iterable[ReachCurve],
+        reach_points: Iterable[ReachPoint],
+        num_clusters: int,
+        N: int,
+    ):
+        """Constructor for GeneralizedMixtureReachSurface.
+
+        Args:
+          reach_curves: A list of ReachCurves to be used in model fitting and
+            reach prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit.
+            This list is of arbitrary length, but it should contain a minimum of
+            p(p-1)/2 points, where p is the number of publishers. Otherwise,
+            there is no guarantee of a unique solution. The reach points
+            represent arbitrary points on the reach surface, involving spend at
+            multiple advertisers
+          num_clusters: number of clusters. Different users in the same cluster
+            have the same probability to receive a new impression.
+          N : A non-negative value that is at least as large as the maximum
+            reach of any of the per-publisher reach functions,
+            e.g., N >= reach_curves[i].max_reach for i=1,2,...,p.
+        """
+
+        self._reach_curves = copy.deepcopy(reach_curves)
+        self._n = len(reach_points)
+        self._p = len(reach_points[0].impressions)
+        self._c = num_clusters
+        self._N = N
+        self._n_iter = 10
+        super().__init__(data=reach_points)
+
+    def update_theta_from_a(self, a, r):
```

Comment:
Define a vectorized function "theta" that computes theta(X), where X is a c x p matrix.  It should return a c x 1 vector representing the evaluation of the function theta at each row of the input X.  Even better would be to define a vectorized function "theta" that maps an n x c x p matrix to an n x c x 1 matrix.  Then, define a second vectorized function "theta_from_a_r_N" that computes the matrix X from a, r and N and then calls theta(X).

Reply:
Thanks Matthew, updated it as best as I can, let me know if this is close to what you have in mind. Would happy to update more.